### PR TITLE
[ie/dailymotion] search support, fix #6126

### LINKED ISSUE #6126: [dailymotion] Support search
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I remove or skip any mandatory\* field

### Checklist

- [X] I'm requesting a site-specific feature
- [X] I've verified that I'm running yt-dlp version **2023.01.06** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

greece

### Example URLs

yt-dlp  --cookies-from-browser firefox -vU --limit-rate 2M --downloader aria2c -N 5  "ytsearch2000:documanters" https://www.dailymotion.com/ 

### Provide a description that is worded well enough to be understood

support to download videos from dailymotion using search string like "ytsearch" option

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
no "dmsearch2000:documanters"  exists as option or similar
```

### PR DESCRIPTION
**Pull request by drzraf**
### Description of your *pull request* and other information

Support retrieving videos from searches

Fixes #6126

Notes:
- Some unrelated tests fail for this extractor, but this is left to #7692
- I'm sorry for the GraphQL blob 🤮. I made some attempts at removing fragments from the requests but DM responded with errors.
- I'm not happy about storing explicitly the search inside info-json `playlist*` keys. I guess this should be avoid (but this is great for naming files). Let me know if this can be avoided/improved.

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details><details><summary>Copilot Summary</summary>  

### Summary
:sparkles::recycle::mag:

Add support for Dailymotion search URLs. Implement a new extractor class `DailymotionSearchIE` that uses the GraphQL API and refactor some common code for obtaining the access token. Register the new extractor in `_extractors.py`.

> _New extractor class_
> _`DailymotionSearchIE`_
> _Uses GraphQL API_

### Walkthrough
*  Add `DailymotionSearchIE` class to handle Dailymotion search URLs ([link](https://github.com/yt-dlp/yt-dlp/pull/8292/files?diff=unified&w=0#diff-780b22dc7eb280f5a7b2bbf79aff17826de88ddcbf2fc1116ba19901827aa4e3R450), [link](https://github.com/yt-dlp/yt-dlp/pull/8292/files?diff=unified&w=0#diff-a0edc55bc9938a7ebebf55833f7fcccf1a5a3050f0a390b40e9da72be9abc17cR695-R725))
   * Define `_VALID_URL` pattern to match search URLs ([link](https://github.com/yt-dlp/yt-dlp/pull/8292/files?diff=unified&w=0#diff-a0edc55bc9938a7ebebf55833f7fcccf1a5a3050f0a390b40e9da72be9abc17cR695-R725))
   * Implement `_real_extract` method to extract search term and delegate to `_fetch_page` method ([link](https://github.com/yt-dlp/yt-dlp/pull/8292/files?diff=unified&w=0#diff-a0edc55bc9938a7ebebf55833f7fcccf1a5a3050f0a390b40e9da72be9abc17cR695-R725))
   * Implement `_fetch_page` method to call `_call_search_api` method and return a playlist of videos ([link](https://github.com/yt-dlp/yt-dlp/pull/8292/files?diff=unified&w=0#diff-a0edc55bc9938a7ebebf55833f7fcccf1a5a3050f0a390b40e9da72be9abc17cR695-R725))
* Refactor access token logic into `_get_token` method in `DailymotionBaseInfoExtractor` class ([link](https://github.com/yt-dlp/yt-dlp/pull/8292/files?diff=unified&w=0#diff-a0edc55bc9938a7ebebf55833f7fcccf1a5a3050f0a390b40e9da72be9abc17cL47-R80))
   * Use `urllib.parse.unquote_plus` to decode search term from URL ([link](https://github.com/yt-dlp/yt-dlp/pull/8292/files?diff=unified&w=0#diff-a0edc55bc9938a7ebebf55833f7fcccf1a5a3050f0a390b40e9da72be9abc17cR4))
   * Call `_get_token` method in `_call_api` and `_call_search_api` methods to avoid code duplication ([link](https://github.com/yt-dlp/yt-dlp/pull/8292/files?diff=unified&w=0#diff-a0edc55bc9938a7ebebf55833f7fcccf1a5a3050f0a390b40e9da72be9abc17cL47-R80), [link](https://github.com/yt-dlp/yt-dlp/pull/8292/files?diff=unified&w=0#diff-a0edc55bc9938a7ebebf55833f7fcccf1a5a3050f0a390b40e9da72be9abc17cL91-R400))
* Add `_call_search_api` method to `DailymotionBaseInfoExtractor` class to perform GraphQL query for search results ([link](https://github.com/yt-dlp/yt-dlp/pull/8292/files?diff=unified&w=0#diff-a0edc55bc9938a7ebebf55833f7fcccf1a5a3050f0a390b40e9da72be9abc17cL91-R400))
   * Use `self._download_json` to send query and return response as JSON object ([link](https://github.com/yt-dlp/yt-dlp/pull/8292/files?diff=unified&w=0#diff-a0edc55bc9938a7ebebf55833f7fcccf1a5a3050f0a390b40e9da72be9abc17cL91-R400))

</details>

### CODE REVIEW

**Comment by drzraf** [comment]
What do you suggest regarding the regexp/URL conflict?

File: yt_dlp/extractor/dailymotion.py [hunk 1]
```diff
@@ -44,35 +45,39 @@ def _real_initialize(self):
...
-                        None, 'Downloading Access Token',
-                        data=urlencode_postdata(data))['access_token']
-                except ExtractorError as e:
-                    if isinstance(e.cause, HTTPError) and e.cause.status == 400:
-                        raise ExtractorError(self._parse_json(
-                            e.cause.response.read().decode(), xid)['error_description'], expected=True)
-                    raise
-                self._set_dailymotion_cookie('access_token' if username else 'client_token', token)
+            token = self._get_token(xid)
             self._HEADERS['Authorization'] = 'Bearer ' + token
```
Comment (seproDev) [inline yt_dlp/extractor/dailymotion.py:81]:
```suggestion
            self._HEADERS['Authorization'] = 'Bearer ' + self._get_token(xid)
```

File: yt_dlp/extractor/dailymotion.py [hunk 2]
```diff
@@ -88,6 +93,310 @@ def _call_api(self, object_type, xid, object_fields, note, filter_extra=None):
             raise ExtractorError(resp['errors'][0]['message'], expected=True)
         return obj
 
+    def _call_search_api(self, term, page, note):
```
Comment (seproDev) [inline yt_dlp/extractor/dailymotion.py:96]: Move in to `DailymotionSearchIE`

File: yt_dlp/extractor/dailymotion.py [hunk 3]
```diff
@@ -88,6 +93,310 @@ def _call_api(self, object_type, xid, object_fields, note, filter_extra=None):
...
+        payload_living_horror = '''
```
Comment (seproDev) [inline yt_dlp/extractor/dailymotion.py:97]: Yeah, this kind of sucks. I think it's probably best to use a slightly minified version on a single line:
```suggestion
        search_query = 'fragment VIDEO_BASE_FRAGMENT on Video { id xid title createdAt stats { id views { id total __typename } __typename } channel { id xid name displayName accountType __typename } duration thumbnailx60: thumbnailURL(size: "x60") thumbnailx120: thumbnailURL(size: "x120") thumbnailx240: thumbnailURL(size: "x240") thumbnailx720: thumbnailURL(size: "x720") aspectRatio __typename } fragment VIDEO_FAVORITES_FRAGMENT on Media { __typename ... on Video { id isInWatchLater __typename } ... on Live { id isInWatchLater __typename } } fragment CHANNEL_BASE_FRAG on Channel { accountType id xid name displayName isFollowed thumbnailx60: logoURL(size: "x60") thumbnailx120: logoURL(size: "x120") thumbnailx240: logoURL(size: "x240") thumbnailx720: logoURL(size: "x720") __typename } fragment PLAYLIST_BASE_FRAG on Collection { id xid name channel { id xid name displayName accountType __typename } description thumbnailx60: thumbnailURL(size: "x60") thumbnailx120: thumbnailURL(size: "x120") thumbnailx240: thumbnailURL(size: "x240") thumbnailx720: thumbnailURL(size: "x720") stats { id videos { id total __typename } __typename } __typename } fragment TOPIC_BASE_FRAG on Topic { id xid name videos(sort: "recent", first: 5) { pageInfo { hasNextPage nextPage __typename } edges { node { id ...VIDEO_BASE_FRAGMENT ...VIDEO_FAVORITES_FRAGMENT __typename } __typename } __typename } stats { id videos { id total __typename } __typename } __typename } query SEARCH_QUERY($query: String!, $shouldIncludeVideos: Boolean!, $shouldIncludeChannels: Boolean!, $shouldIncludePlaylists: Boolean!, $shouldIncludeTopics: Boolean!, $shouldIncludeLives: Boolean!, $page: Int, $limit: Int, $sortByVideos: SearchVideoSort, $durationMinVideos: Int, $durationMaxVideos: Int, $createdAfterVideos: DateTime) { search { id videos( query: $query first: $limit page: $page sort: $sortByVideos durationMin: $durationMinVideos durationMax: $durationMaxVideos createdAfter: $createdAfterVideos ) @include(if: $shouldIncludeVideos) { metadata { algorithm { uuid __typename } __typename } pageInfo { hasNextPage nextPage __typename } totalCount edges { node { id ...VIDEO_BASE_FRAGMENT ...VIDEO_FAVORITES_FRAGMENT __typename } __typename } __typename } hasLives: lives(query: $query, first: $limit, page: $page) { totalCount __typename } lives(query: $query, first: $limit, page: $page) @include(if: $shouldIncludeLives) { metadata { algorithm { uuid __typename } __typename } pageInfo { hasNextPage nextPage __typename } totalCount edges { node { id xid title thumbnail: thumbnailURL(size: "x240") thumbnailx60: thumbnailURL(size: "x60") thumbnailx120: thumbnailURL(size: "x120") thumbnailx240: thumbnailURL(size: "x240") thumbnailx720: thumbnailURL(size: "x720") audienceCount aspectRatio isOnAir channel { id xid name displayName accountType __typename } __typename } __typename } __typename } channels(query: $query, first: $limit, page: $page) @include(if: $shouldIncludeChannels) { metadata { algorithm { uuid __typename } __typename } pageInfo { hasNextPage nextPage __typename } totalCount edges { node { id ...CHANNEL_BASE_FRAG __typename } __typename } __typename } playlists: collections(query: $query, first: $limit, page: $page) @include(if: $shouldIncludePlaylists) { metadata { algorithm { uuid __typename } __typename } pageInfo { hasNextPage nextPage __typename } totalCount edges { node { id ...PLAYLIST_BASE_FRAG __typename } __typename } __typename } topics(query: $query, first: $limit, page: $page) @include(if: $shouldIncludeTopics) { metadata { algorithm { uuid __typename } __typename } pageInfo { hasNextPage nextPage __typename } totalCount edges { node { id ...TOPIC_BASE_FRAG __typename } __typename } __typename } __typename } }'
```

File: yt_dlp/extractor/dailymotion.py [hunk 4]
```diff
@@ -88,6 +93,310 @@ def _call_api(self, object_type, xid, object_fields, note, filter_extra=None):
...
+      __typename
+    }
+    __typename
+  }
+}
+'''
+
+        if not self._HEADERS.get('Authorization'):
+            token = self._get_token(term)
+            self._HEADERS['Authorization'] = 'Bearer ' + token
```
Comment (seproDev) [inline yt_dlp/extractor/dailymotion.py:379]:
```suggestion
            self._HEADERS['Authorization'] = 'Bearer ' + self._get_token(term)
```

File: yt_dlp/extractor/dailymotion.py [hunk 5]
```diff
@@ -88,6 +93,310 @@ def _call_api(self, object_type, xid, object_fields, note, filter_extra=None):
...
+                    'shouldIncludeChannels': False,
+                    'shouldIncludeLives': False,
+                    'shouldIncludePlaylists': False,
+                    'shouldIncludeTopics': False,
+                    'shouldIncludeVideos': True
+                }
+            }).encode(), headers=self._HEADERS)
+        obj = resp['data']['search']
+        if not obj:
+            raise ExtractorError(resp['errors'][0]['message'], expected=True)
```
Comment (seproDev) [inline yt_dlp/extractor/dailymotion.py:397]: Should be non-fatal:
```suggestion
        obj = traverse_obj(resp, ('data', 'search'))
        if not obj:
            raise ExtractorError(traverse_obj(resp, ('errors', 0, 'message'))
                                 or 'Could not fetch search data', expected=True)
```

File: yt_dlp/extractor/dailymotion.py [hunk 6]
```diff
@@ -383,6 +692,37 @@ def _extract_embed_urls(cls, url, webpage):
...
+    def _real_extract(self, url):
+        term = urllib.parse.unquote_plus(self._match_id(url))
+        entries = OnDemandPagedList(functools.partial(
+            self._fetch_page, term), self._PAGE_SIZE)
+        return self.playlist_result(entries, term, 'Dailymotion search of %s' % term)
+
+
 class DailymotionUserIE(DailymotionPlaylistBaseIE):
     IE_NAME = 'dailymotion:user'
     _VALID_URL = r'https?://(?:www\.)?dailymotion\.[a-z]{2,3}/(?!(?:embed|swf|#|video|playlist)/)(?:(?:old/)?user/)?(?P<id>[^/]+)'
```
Comment (seproDev) [inline yt_dlp/extractor/dailymotion.py:728]: The search URLs need to be excluded from `DailymotionUserIE`
```suggestion
    _VALID_URL = r'https?://(?:www\.)?dailymotion\.[a-z]{2,3}/(?!(?:embed|swf|#|video|playlist|search)/)(?:(?:old/)?user/)?(?P<id>[^/]+)'
```

File: yt_dlp/extractor/dailymotion.py [hunk 7]
```diff
@@ -383,6 +692,37 @@ def _extract_embed_urls(cls, url, webpage):
...
+                'https://www.dailymotion.com/video/' + node['xid'],
+                DailymotionIE.ie_key(),
+                node['xid']
+            )
+
+    def _real_extract(self, url):
+        term = urllib.parse.unquote_plus(self._match_id(url))
+        entries = OnDemandPagedList(functools.partial(
+            self._fetch_page, term), self._PAGE_SIZE)
+        return self.playlist_result(entries, term, 'Dailymotion search of %s' % term)
```
Comment (seproDev) [inline yt_dlp/extractor/dailymotion.py:723]:
```suggestion
        return self.playlist_result(
            OnDemandPagedList(functools.partial(self._fetch_page, term), self._PAGE_SIZE), term, term)
```

File: yt_dlp/extractor/dailymotion.py [hunk 8]
```diff
@@ -383,6 +692,37 @@ def _extract_embed_urls(cls, url, webpage):
...
+    def _fetch_page(self, term, page):
+        page += 1
+        videos = self._call_search_api(term, page, 'Searching "%s", page %d' % (term, page))['videos']
+        for edge in videos['edges']:
+            node = edge['node']
+            yield self.url_result(
+                'https://www.dailymotion.com/video/' + node['xid'],
+                DailymotionIE.ie_key(),
+                node['xid']
+            )
```
Comment (seproDev) [inline yt_dlp/extractor/dailymotion.py:717]: Should be non-fatal:
```suggestion
        response = self._call_search_api(term, page, f'Searching "{term}" page {page}')
        for edge in traverse_obj(response, ('videos', 'edges', lambda _, x: x['node']['xid'])) or []:
            xid = edge['node']['xid']
            yield self.url_result(f'https://www.dailymotion.com/video/{xid}', DailymotionIE, xid)
```

**Review by seproDev — CHANGES_REQUESTED** [review]

**Comment by pukkandan** [comment]
> Ah, looking at some other extractors, search functionality should be dealt with by having an InfoExtractor with `_SEARCH_KEY` and `_search_results` defined. You can then add an additional `SearchURLIE` that matches the search URLs and calls your `_search_results` function. Take a look at the YouTube extractor to see what I mean.

This is not necessary. Having a search URL IE without a SearchIE is fine

**Review by seproDev — APPROVED** [review]

File: yt_dlp/extractor/dailymotion.py [hunk 9]
```diff
@@ -383,9 +387,61 @@ def _extract_embed_urls(cls, url, webpage):
...
...e thumbnail: thumbnailURL(size: "x240") thumbnailx60: thumbnailURL(size: "x60") thumbnailx120: thumbnailURL(size: "x120") thumbnailx240: thumbnailURL(size: "x240") thumbnailx720: thumbnailURL(size: "x720") audienceCount aspectRatio isOnAir channel { id xid name displayName accountType __typename } __typename } __typename } __typename } channels(query: $query, first: $limit, page: $page) @include(if: $shouldIncludeChannels) { metadata { algorithm { uuid __typename } __typename } pageInfo { hasNextPage nextPage __typename } totalCount edges { node { id ...CHANNEL_BASE_FRAG __typename } __typename } __typename } playlists: collections(query: $query, first: $limit, page: $page) @include(if: $shouldIncludePlaylists) { metadata { algorithm { uuid __typename } __typename } pageInfo { hasNextPage nextPage __typename } totalCount edges { node { id ...PLAYLIST_BASE_FRAG __typename } __typename } __typename } topics(query: $query, first: $limit, page: $page) @include(if: $shouldIncludeTopics) { metadata { algorithm { uuid __typename } __typename } pageInfo { hasNextPage nextPage __typename } totalCount edges { node { id ...TOPIC_BASE_FRAG __typename } __typename } __typename } __typename } }'
```
Comment (bashonly) [inline yt_dlp/extractor/dailymotion.py:404]: Does the request still work if we remove any of this? We are only extracting the IDs, we don't need thumbnail etc

File: yt_dlp/extractor/dailymotion.py [hunk 10]
```diff
@@ -383,9 +387,61 @@ def _extract_embed_urls(cls, url, webpage):
...
+                    'shouldIncludeLives': False,
+                    'shouldIncludePlaylists': False,
+                    'shouldIncludeTopics': False,
+                    'shouldIncludeVideos': True
+                }
+            }).encode(), headers=self._HEADERS)
+        obj = traverse_obj(resp, ('data', 'search', {dict}))
+        if not obj:
+            raise ExtractorError(traverse_obj(resp, ('errors', 0, 'message'))
+                                 or 'Could not fetch search data', expected=True)
```
Comment (bashonly) [inline yt_dlp/extractor/dailymotion.py:426]:
```suggestion
            raise ExtractorError(
                traverse_obj(resp, ('errors', 0, 'message', {str})) or 'Could not fetch search data', expected=True)
```
Why is this expected, though? (I'm just asking a question, not necessarily suggesting anything other than the formatting be changed)

File: yt_dlp/extractor/dailymotion.py [hunk 11]
```diff
@@ -383,9 +387,61 @@ def _extract_embed_urls(cls, url, webpage):
...
+        return obj
+
+    def _fetch_page(self, term, page):
+        page += 1
+        response = self._call_search_api(term, page, f'Searching "{term}" page {page}')
+        for edge in traverse_obj(response, ('videos', 'edges', lambda _, x: x['node']['xid'])) or []:
+            xid = edge['node']['xid']
```
Comment (bashonly) [inline yt_dlp/extractor/dailymotion.py:433]:
```suggestion
        for xid in traverse_obj(response, ('videos', 'edges', ..., 'node', 'xid')):
```

**Review by bashonly — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/dailymotion.py (hunk 10 above)
Comment (seproDev) [reply yt_dlp/extractor/dailymotion.py]: Yes, I think this shouldn't be expected. Empty search results do not raise this error.

File: yt_dlp/extractor/dailymotion.py [hunk 12]
```diff
@@ -393,9 +397,56 @@ def _extract_embed_urls(cls, url, webpage):
...
+    def _real_extract(self, url):
+        term = urllib.parse.unquote_plus(self._match_id(url))
+        return self.playlist_result(
+            OnDemandPagedList(functools.partial(self._fetch_page, term), self._PAGE_SIZE), term, term)
+
+
 class DailymotionUserIE(DailymotionPlaylistBaseIE):
     IE_NAME = 'dailymotion:user'
-    _VALID_URL = r'https?://(?:www\.)?dailymotion\.[a-z]{2,3}/(?!(?:embed|swf|#|video|playlist)/)(?:(?:old/)?user/)?(?P<id>[^/]+)'
+    _VALID_URL = r'https?://(?:www\.)?dailymotion\.[a-z]{2,3}/(?!(?:embed|swf|#|video|playlist|search)/)(?:(?:old/)?user/)?(?P<id>[^/]+)'
```
Comment (bashonly) [inline yt_dlp/extractor/dailymotion.py:449]:
```suggestion
    _VALID_URL = r'https?://(?:www\.)?dailymotion\.[a-z]{2,3}/(?!(?:embed|swf|#|video|playlist|search)/)(?:(?:old/)?user/)?(?P<id>[^/?#]+)'
```

**Review by bashonly — APPROVED** [review]

**Pull request merged**